radix-engine/src/kernel/kernel.rs

```diff
@@ -365,44 +465,101 @@ where
+                M::on_move_module(api, MoveModuleEvent::StoreAccess(&store_access))
+            },
+        };
 
-        M::after_move_modules(
-            src_node_id,
-            src_partition_number,
-            dest_node_id,
-            dest_partition_number,
-            &store_access,
-            self,
-        )?;
+        self.current_frame
+            .move_module(
+                src_node_id,
+                src_partition_number,
+                dest_node_id,
+                dest_partition_number,
+                &mut self.heap,
+                self.store,
+                &mut handler,
+            )
+            .map_err(|e| match e {
+                CallbackError::Error(e) => RuntimeError::KernelError(KernelError::CallFrameError(
+                    CallFrameError::MoveModuleError(e),
+                )),
+                CallbackError::CallbackError(e) => e,
+            })?;
 
         Ok(())
     }
 }
 
+// TODO: Remove
 impl<'g, M, S> KernelInternalApi<M> for Kernel<'g, M, S>
 where
     M: KernelCallbackObject,
     S: SubstateStore,
 {
-    fn kernel_get_node_visibility(&self, node_id: &NodeId) -> NodeVisibility {
+    fn kernel_get_node_visibility(&mut self, node_id: &NodeId) -> NodeVisibility {
+        let mut read_only = as_read_only!(self);
+        read_only.kernel_get_node_visibility(node_id)
+    }
+
+    fn kernel_get_current_depth(&mut self) -> usize {
+        let mut read_only = as_read_only!(self);
+        read_only.kernel_get_current_depth()
+    }
+
+    fn kernel_get_system_state(&mut self) -> SystemState<'_, M> {
+        let caller_actor = match self.prev_frame_stack.last() {
+            Some(call_frame) => call_frame.actor(),
+            None => {
+                // This will only occur on initialization
+                self.current_frame.actor()
+            }
+        };
+        SystemState {
+            system: &mut self.callback,
+            current_actor: self.current_frame.actor(),
+            caller_actor,
+        }
+    }
+
+    fn kernel_read_bucket(&mut self, bucket_id: &NodeId) -> Option<BucketSnapshot> {
+        let mut read_only = as_read_only!(self);
+        read_only.kernel_read_bucket(bucket_id)
+    }
+
+    fn kernel_read_proof(&mut self, proof_id: &NodeId) -> Option<ProofSnapshot> {
+        let mut read_only = as_read_only!(self);
+        read_only.kernel_read_proof(proof_id)
+    }
+}
+
+struct KernelReadOnly<'g, M>
+where
+    M: KernelCallbackObject,
+{
+    current_frame: &'g CallFrame<M::LockData>,
+    prev_frame: Option<&'g CallFrame<M::LockData>>,
+    heap: &'g Heap,
+    callback: &'g mut M,
+}
+
+impl<'g, M> KernelInternalApi<M> for KernelReadOnly<'g, M>
+where
+    M: KernelCallbackObject,
+{
+    fn kernel_get_node_visibility(&mut self, node_id: &NodeId) -> NodeVisibility {
         self.current_frame.get_node_visibility(node_id)
     }
 
-    fn kernel_get_current_depth(&self) -> usize {
+    fn kernel_get_current_depth(&mut self) -> usize {
```

Comment:
Why `&mut`

Reply:
This is a bit of a hack right now, but required as the `Kernel` impl of `KernelInternalApi` wraps `KernelReadOnly` and this requires `mut` but I think I can just directly implement these for these cases.